coding-style.md

```diff
@@ -0,0 +1,552 @@
+### Template Metaprogramming
+
+- Use the `typename` keyword only to access dependent types. For general
+  template parameters, use `class` instead:
+
+  ```cpp
+  template <class T>
+  struct foo {
+    using type = typename T::type;
+  };
+  ```
+
+- Use `T` for generic, unconstrained template parameters and `x` for generic
+  function arguments. Suffix both with `s` for template parameter packs:
+
+  ```cpp
+  template <class T, class... Ts>
+  auto f(T x, Ts... xs) {
+    // ...
+  }
+  ```
+
+- Break `using name = ...` statements always directly after `=` if they do not
+  fit in one line.
+
+- Use one level of indentation per "open" template and place the closing `>`,
+  `>::type` or `>::value` on its own line. For example:
+  ```cpp
+  using optional_result_type =
+    typename std::conditional<
+      std::is_same<result_type, void>::value,
+      bool,
+      optional<result_type>
+    >::type;
+  ```
+
+- When dealing with "ordinary" templates, use indentation based on the position
+  of the last opening `<`:
+  ```cpp
+  using type = quite_a_long_template_which_needs_a_break<std::string,
+                                                         double>;
+  ```
+
+- When adding new type traits, also provide `*_t` and/or `*_v` helpers:
+  ```cpp
+  template <class T>
+  using my_trait_t = typename my_trait<T>::type;
+
+  template <class T>
+  constexpr auto my_trait_v = my_trait<T>::value;
+  ```
+
+### Logging
+
+- Available log levels are *ERROR*, *WARNING*, *INFO*, *DEBUG* and *TRACE*.
+
+- Messages can be sent by using the `VAST_\<level>` and `VAST_\<level>_ANON`
+  macros.
+
+- The `VAST_\<level>` macros are intended to be used with a subject as the fist
+  argument. Use them to create a sentence of the form 'subject verb object'.
+  For example:
+  ```cpp
+  VAST_WARNING(self, "got a request for unknown query ID", query_id);
+  ```
+
+  Which gets printed as:
+  ```sh
+  index got a request for unkown query ID <181f5b1f-c673-4e98-acd2-ef762fe567a2>
+  ```
+
+- The `VAST_\<level>_ANON` macros are available for situations where **n**o
+  **s**ubject is suitable for the first argument to the other macro variant.
+  They can either be used to inject a subject manually, or create a message
+  in the *gerund* form. For example:
+  ```cpp
+  VAST_DEBUG_ANON(__func__, "creating directory", dir);
+  ```
+
+- By default, use the `VAST_\<level>` variants.
+
+- Try to restrict usage of the `VAST_INFO` message type to the main actors.
+  Info is the chattiest level that most users will see, so it should require
+  no or only little understanding of VASTs system architecture for the
+  reader to understand.
+
+- Messages sent at the trace level add the additional effect of writing a
+  second message at the exit of the current scope. The trace level can be
+  used to create a trace of the call stack with fine grained control over
+  its depth. Omit trace messages from helper functions and generic / general
+  purpose algorithm implementations.
```

Comment:
This was moved into a separate VAST_TRACE_SCOPE macro.

Reply:
Can you make a rephrasing suggestion? I don't know how this works.